### ISSUE TITLE
napari workflow: Cryptic error message when attempting to run a workflow on cpu-low / with not enough memory

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/fractal-tasks-core#256]
I've tried to run a measurement workflow on a large well_ROI (8404 x 8114 image). When running with cpu-low (new 1 GB of memory), this runs into a cryptic napari workflow error where it's not able to start a new thread:

```
2022-12-07 10:42:10,331; INFO; ROI 1/1: region=(slice(0, 1, None), slice(0, 8404, None), slice(0, 8114, None))
Traceback (most recent call last):
  File "/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/fractal_tasks_core/napari_workflows_wrapper.py", line 604, in <module>
    run_fractal_task(
  File "/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/fractal_tasks_core/_utils.py", line 89, in run_fractal_task
    metadata_update = task_function(**task_args.dict())
  File "/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/fractal_tasks_core/napari_workflows_wrapper.py", line 484, in napari_workflows_wrapper
    outputs = wf.get(list_outputs)
  File "/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/napari_workflows/_workflow.py", line 80, in get
    return dask_get(self._tasks, name)
  File "/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/dask/threaded.py", line 89, in get
    results = get_async(
  File "/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/dask/local.py", line 511, in get_async
    raise_exception(exc, tb)
  File "/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/dask/local.py", line 319, in reraise
    raise exc
  File "/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/dask/local.py", line 224, in execute_task
    result = _execute_task(task, data)
  File "/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/dask/core.py", line 119, in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
  File "/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/napari_skimage_regionprops/_regionprops.py", line 103, in regionprops_table
    table = sk_regionprops_table(np.asarray(labels).astype(int), intensity_image=np.asarray(image),
  File "/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/dask/array/core.py", line 1692, in __array__
    x = self.compute()
  File "/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/dask/base.py", line 315, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/dask/base.py", line 598, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/dask/threaded.py", line 89, in get
    results = get_async(
  File "/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/dask/local.py", line 499, in get_async
    fire_tasks(chunksize)
  File "/data/homes/fractal/joel/fractal_v1/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.6.2/venv/lib/python3.8/site-packages/dask/local.py", line 494, in fire_tasks
    fut = submit(batch_execute_tasks, each_args)
  File "/data/homes/fractal/.conda/envs/fractal-server-v1-b3/lib/python3.8/concurrent/futures/thread.py", line 188, in submit
    self._adjust_thread_count()
  File "/data/homes/fractal/.conda/envs/fractal-server-v1-b3/lib/python3.8/concurrent/futures/thread.py", line 213, in _adjust_thread_count
    t.start()
  File "/data/homes/fractal/.conda/envs/fractal-server-v1-b3/lib/python3.8/threading.py", line 852, in start
    _start_new_thread(self._bootstrap, ())
RuntimeError: can't start new thread

```

I'll test a bit further to see if this affects all workflows (i.e. 1 GB is never enough, would be weird) or if it's a cryptic message when there is not enough memory for a given workflow with a large ROI image.

**Comment by jluethi** [comment]
Ok, can now confirm that this also happens on the tiny dataset, so it's not dependent on the size of the image. It appears that napari workflows can't run on a 1 GB memory node.

I'll report it to napari workflows to see if we can improve the error message (/ to make sure this is documented). In the meantime: Why did we switch CPU-low to 1GB memory? Let's have a brief discussion next week about managing executors/task requirements :)

**Comment by tcompa** [comment]
> In the meantime: Why did we switch CPU-low to 1GB memory?

By mistake. I've moved back to 7G in the demos repository.

**Comment by jluethi** [comment]
Hmm, interesting. This was run on a slurm job with a single CPU. Would that still get the config settings of the whole node?
With our lowest executor having 7 GBs of memory, I've never seen this come up. But good to be aware of where we'd start modifying some of those settings :)

**Comment by tcompa** [comment]
If it's not too cumbersome, could you also test this with the 7G memory requirement?

Also: which cluster are you using?
At the UZH one, the memory requirement is not a hard limit - so that I'd be confused if we see such a different behavior between _requested_ 1G and _requested_ 7G cases. I guess it's rather the number of threads which matters here, and not the memory.

**Comment by jluethi** [comment]
This ran at UZH. I'll rerun it with the 7GB cpu-low executor after the next server restart :)

It ran at UZH with the CPU-mid executor, which also only requested 1 CPU core (not sure why this changed) and 15 GB of memory. But it reliably produces the error above when running on the 1 GB memory cpu-low executor

**Comment by tcompa** [comment]
A quite useful test would be to include
```
export OMP_NUM_THREADS=1
export MKL_NUM_THREADS=1
export OPENBLAS_NUM_THREADS=1
```
in your `WORKER_INIT` variable.

Note that those variables were already there in some very old version of Fractal, although in that case it was due to parsl repeated imports opening too many threads. I guess a similar issue could be the problem here.

**Comment by tcompa** [comment]
A couple more references (just for the record):
* https://github.com/dask/dask/issues/4240
* https://github.com/SciTools/iris/issues/3237

**Comment by jluethi** [comment]
Closing, as with the new init and current memory defaults, this didn't come up as an issue anymore.

**Comment by jluethi** [comment]
I've been running tests where I tried lower amounts of memory again, because memory usage appeared to be quite low. Lowered the napari-workflow memory requirement to 2 GB & running on 1 CPU. I then ran into the error above again.
Increasing CPUs had no effect, increasing the memory from 2 to 4 GB made this error disappear (and I then ran into a an actual `numpy.core._exceptions.MemoryError`.
Increasing it back to 8GB lead to

I'm a bit surprised at that, given that it was just processing a single FOV in 2D (shouldn't require 8GB of memory). Also, the sacct in slurm only showed memory usage in the MBs (but I guess I don't really trust this statistic now).
